spin1_api/spin1_api.c

```diff
@@ -1551,19 +1553,32 @@ uint spin1_schedule_callback(callback_t cback, uint arg0, uint arg1,
 */
 uint spin1_trigger_user_event(uint arg0, uint arg1)
 {
-    if (!user_pending) {
-        /* remember callback arguments */
-        user_arg0 = arg0;
-        user_arg1 = arg1;
-        user_pending = TRUE;
+    uint cpsr = spin1_int_disable();
 
-        /* trigger software interrupt in the VIC */
-        vic[VIC_SOFT_SET] = (1 << SOFTWARE_INT);
+    uint new_end = (user_event_queue.end + 1) % USER_EVENT_QUEUE_SIZE;
+
+    if (new_end == user_event_queue.start) {
+        spin1_mode_restore(cpsr);
+
+        #if (API_DIAGNOSTICS == TRUE)
+        diagnostics.user_event_queue_full++;
+        #endif // (API_DIAGNOSTICS == TRUE)
 
-        return (SUCCESS);
-    } else {
         return (FAILURE);
     }
+
+    user_event_queue.queue[user_event_queue.end].arg0 = arg0;
+    user_event_queue.queue[user_event_queue.end].arg1 = arg1;
+
+    /* if no other user event, trigger on now */
+    if (user_event_queue.start == user_event_queue.end) {
```

Comment:
I wouldn't bother checking if the queue is empty. It is correct to enable the software interrupt even if the queue was not full.